### ISSUE TITLE
`calicoctl status` giving `undefined symbol: rl_signal_event_hook` under certain bash versions

### ISSUE THREAD

**Issue by erandu** [projectcalico/calico-containers#1063]
I have follow this guide https://github.com/projectcalico/calico-containers/tree/master/docs/calico-with-docker/docker-network-plugin with vagrant coreOS installation and I have a problem when I execute sudo calicoctl status,  I have an error message "Couldn't connect to bird." My other problem is the error message "sh: symbol lookup error: sh: undefined symbol: rl_signal_event_hook" and I don't understand what is the origin of this error message. 

So when I execute sudo calicoctl status, I have this :

```
core@calico-01 ~ $ sudo calicoctl status
sh: symbol lookup error: sh: undefined symbol: rl_signal_event_hook
sh: symbol lookup error: sh: undefined symbol: rl_signal_event_hook
sh: symbol lookup error: sh: undefined symbol: rl_signal_event_hook
sh: symbol lookup error: sh: undefined symbol: rl_signal_event_hook
calico-node container is running. Status: Up 43 hours
Running felix version 1.4.0rc2

IPv4 BGP status
IP: 172.17.8.101    AS Number: 64511 (inherited)
/bin/sh: symbol lookup error: /bin/sh: undefined symbol: rl_signal_event_hook
Couldn't connect to bird.
IPv6 BGP status
No IPv6 address configured.

core@calico-01 ~ $ 
```

~~I am on Fedora.~~  [SMC] User confirmed they were using a CoreOS alpha VM started with Vagrant.

calicoctl: **0.20.0-dev**, the vagrant script downloads that from: http://www.projectcalico.org/builds/calicoctl 
OS: CoreOS alpha 1097.0.0

**Comment by fasaxc** [comment]
These issue look related: 

https://github.com/projectcalico/calico-containers/issues/60 looks like we tracked it down to a bad Python build @teemow managed to get a build without the error; @tomdee did we ever upstream his fix?

https://github.com/docker/compose/issues/170 Same problem with docker compose (they did manage to fix but thread is long and I didn't see a reference to the fix).

If compiling with  a newer Python would help, we have a build container image in https://github.com/projectcalico/calico that builds Python 2.7 from source on SL6.5.  So that uses latest Python with old glibc.

**Comment by fasaxc** [comment]
@erandu tried downloading the 0.20.0 release version but that shows the same error as 0.20.0-dev.

**Comment by damhau** [comment]
Hi Guys,

In the meantime you can use the docker version of calicoctl with the following command:
/usr/bin/docker run --net=host --privileged -v /var/run/docker.sock:/var/run/docker.sock -v /var/run/calico/bird.ctl:/var/run/calico/bird.ctl calico/ctl status

Tested on Coreos 1109.1.0

Cheers,

Damien

**Comment by damhau** [comment]
of course replace status with the calicoctl subcommand you want to run.

**Comment by damhau** [comment]
and for the final touch add an alias and you re good to go:
alias calicoctl='/usr/bin/docker run --net=host --privileged -v /var/run/docker.sock:/var/run/docker.sock -v /var/run/calico/bird.ctl:/var/run/calico/bird.ctl calico/ctl'

**Comment by tomdee** [comment]
We're building with a really old version of glibc to avoid this kind of issue so I'm not sure what's up
```
 docker run --rm calico/build ldd --version
ldd (Debian EGLIBC 2.13-38+deb7u8) 2.13
```

**Comment by erandu** [comment]
@damhau it works fine with your following command, but only on my node 1.
On node 2, I have this : 
core@calico-02 ~ $ /usr/bin/docker run --net=host --privileged -v /var/run/docker.sock:/var/run/docker.sock -v /var/run/calico/bird.ctl:/var/run/calico/bird.ctl calico/ctl status
Unable to connect to server control socket (/var/run/calico/bird6.ctl): No such file or directory
sh: write error: Broken pipe
calico-node container is running. Status: Up 10 minutes
Running felix version 1.4.0rc2
Error connecting to etcd.  Ensure ETCD_ENDPOINTS or ETCD_AUTHORITY is set properly.

IPv4 BGP status
IP: unknown    AS Number: unknown
+--------------+-------------------+-------+----------+-------------+
| Peer address |     Peer type     | State |  Since   |     Info    |
+--------------+-------------------+-------+----------+-------------+
| 172.17.8.101 | node-to-node mesh |   up  | 13:32:29 | Established |
| 172.17.8.103 | node-to-node mesh |   up  | 13:32:25 | Established |
+--------------+-------------------+-------+----------+-------------+

IPv6 BGP status
IP: unknown    AS Number: unknown
Couldn't connect to bird.


I think it's because my etcd only run on node 1, do you have a solution for node 2?

**Comment by fasaxc** [comment]
I think @hector2jf has found the cause of the symbol error under #1075:

> 
> Hello there,
> 
> I am using CoreOS `1068.8.0` which uses `bash` version `4.3` on which there is a bug https://bugs.archlinux.org/task/39533. This seems to affect to the output of some commands of `calicoctl`. Regardless of this problem, calico seems to be working fine.
> 
> 
> ```
> $  sudo ETCD_AUTHORITY=172.17.8.102:2379 calicoctl pool show
> sh: symbol lookup error: sh: undefined symbol: rl_signal_event_hook
> sh: symbol lookup error: sh: undefined symbol: rl_signal_event_hook
> sh: symbol lookup error: sh: undefined symbol: rl_signal_event_hook
> sh: symbol lookup error: sh: undefined symbol: rl_signal_event_hook
> +----------------+-------------------+
> |   IPv4 CIDR    |      Options      |
> +----------------+-------------------+
> | 192.168.0.0/16 | ipip,nat-outgoing |
> +----------------+-------------------+
> +--------------------------+---------+
> |        IPv6 CIDR         | Options |
> +--------------------------+---------+
> | fd80:24e2:f998:72d6::/64 |         |
> +--------------------------+---------+
> ```
> ```
> $ sudo ETCD_AUTHORITY=172.17.8.103:2379 calicoctl status
> sh: symbol lookup error: sh: undefined symbol: rl_signal_event_hook
> sh: symbol lookup error: sh: undefined symbol: rl_signal_event_hook
> sh: symbol lookup error: sh: undefined symbol: rl_signal_event_hook
> sh: symbol lookup error: sh: undefined symbol: rl_signal_event_hook
> calico-node container is running. Status: Up 2 hours
> Running felix version 1.4.0rc2
> 
> IPv4 BGP status
> IP: 10.1.3.2    AS Number: 64511 (inherited)
> /bin/sh: symbol lookup error: /bin/sh: undefined symbol: rl_signal_event_hook
> Couldn't connect to bird.
> IPv6 BGP status
> No IPv6 address configured.```
> 
> 
> I know this issue is related to the bug in bash rather than to you but I wanted to let you know.

**Comment by dm0-** [comment]
Copying my response here from over at coreos/bugs#1477...

What seems to be happening is that the `calicoctl` binary from http://www.projectcalico.org/builds/calicoctl extracts a bunch of libraries including libreadline to a temporary directory (e.g. `/tmp/_MEIpFhcZH/libreadline.so.6`) as a method of bundling Python into a single executable.

Our bash program is linked against the version under `/lib64` which defines `rl_signal_event_hook`, while the libreadline extracted into `/tmp` does not define it.  While `calicoctl` is executing, it sets the `LD_LIBRARY_PATH` environment variable, causing subprocesses to load the libraries found in `/tmp`.  The dynamic linker then prints those error messages when it can't resolve the symbol.

The reason other distros like Debian and Fedora don't have this error is because they have a copy of readline built into bash instead of linking it with the installed system library.  (And Debian's `/bin/sh` is symlinked to `dash`, which isn't linked to libreadline either.)

**Comment by matthewdupre** [comment]
I had a little look for the docker commit that fixed the issue for them.  I'm not fully confident, but I think it might have been this one: https://github.com/docker/compose/pull/1474.

Reports also suggest it could be fixed by building on a much newer system, but that will likely introduce glibc incompatibilities with (e.g.) CentOS 6.

**Comment by fasaxc** [comment]
Pretty sure Rob is working on this.

**Comment by robbrockbank** [comment]
This is resolved with the latest release.

**Comment by PranjalMishra** [comment]
Running into this issue on DCOS cluster. Using the universe package of Calico 0.1.0. Is this issue fixed there?

**Comment by caseydavenport** [comment]
@PranjalMishra - this should be fixed in version v0.21.0.  Could you make sure you're using that version (a la `calicoctl version`)?

**Comment by PranjalMishra** [comment]
@caseydavenport - thanks for the response. In the DCOS cluster I have following repos,
https://universe.mesosphere.com/repo
and
https://universe.mesosphere.com/repo-1.7
Both of these have Calico package which is older.

How to get v0.21.0?

**Comment by caseydavenport** [comment]
@djosborne ^

**Comment by timstoop** [comment]
Latest version still has this error on CoreOS stable:

```
core@ip-10-100-4-70 ~ $ sudo calicoctl version
sh: symbol lookup error: sh: undefined symbol: rl_signal_event_hook
sh: symbol lookup error: sh: undefined symbol: rl_signal_event_hook
sh: symbol lookup error: sh: undefined symbol: rl_signal_event_hook
sh: symbol lookup error: sh: undefined symbol: rl_signal_event_hook
0.21.0-dev
```

**Comment by djosborne** [comment]
@PranjalMishra Calico's DC/OS v0.2.0 was cut over the weekend. It includes node v0.21.0

**Comment by TomBlossom** [comment]
@djosborne as discussed - on 0.22.0-dev I had this issue while following the guide 

http://docs.projectcalico.org/v1.5/getting-started/docker/installation/vagrant-coreos/
http://docs.projectcalico.org/v1.5/getting-started/docker/tutorials/basic

```
core@calico-02 ~ $ cat /etc/lsb-release 
DISTRIB_ID=CoreOS
DISTRIB_RELEASE=1122.3.0
DISTRIB_CODENAME="MoreOS"
DISTRIB_DESCRIPTION="CoreOS 1122.3.0 (MoreOS)"
```

**Comment by robbrockbank** [comment]
@TomBlossom does this just affect the current master build, or does it also not work on the latest release of the calicoctl?

**Comment by caseydavenport** [comment]
Should be fixed in v1.0.0-beta.
